### PR TITLE
feat(Select, ComboBox, Option): leverage context instead of cloning React elements

### PR DESCRIPTION
**Pull request by nikgraf**
@jpuri can you have a closer look? I think this should improve the performance as we can avoid cloning elements here.

### CODE REVIEW

File: src/__tests__/Option-test.js [hunk 1]
```diff
@@ -10,41 +10,52 @@ import TestUtils from 'react-addons-test-utils';
...
+    shallowRenderer.render(
+      <Option value="rome" data-custom="example">Rome</Option>,
+      { isDisabled: false, }
     );
+    const option = shallowRenderer.getRenderOutput();
 
-    const div = TestUtils.findRenderedDOMComponentWithTag(option, 'div');
-    expect(div.getAttribute('data-custom')).toBe('example');
+    expect(option.props['data-custom']).toBe('example');
   });
```
Comment (jpuri) [inline src/__tests__/Option-test.js]: Hey @nikgraf ,
Any specific advantage of using shallow rendering here ?

File: src/components/ComboBox.js [hunk 2]
```diff
@@ -729,7 +745,7 @@ export default class ComboBox extends Component {
                     onMouseLeave={ ::this._onMouseLeaveAtOption }
                     onMouseDown={ (event) => event.preventDefault() }
                     role="option">
-                  { option }
+                  { entry }
                 </li>
```
Comment (jpuri) [inline src/components/ComboBox.js]: :+1:

File: src/components/Option.js [hunk 3]
```diff
@@ -53,10 +49,17 @@ export default class Option extends Component {
...
+      PropTypes.string,
+      PropTypes.number,
+    ]),
+  };
+
   static defaultProps = {
-    _isHovered: false,
     _isDisplayedAsSelected: false,
-    _isDisabled: false,
   };
```
Comment (jpuri) [inline src/components/Option.js]: How about using just 'hovered' or 'disabled' like html and getting rid of prefix 'is'.

File: src/__tests__/Option-test.js (hunk 1 above)
Comment (nikgraf) [inline src/__tests__/Option-test.js]: We can inject context.

File: src/components/Option.js (hunk 3 above)
Comment (nikgraf) [inline src/components/Option.js]: `disabled` sounds good - wanna do it?

**Pull request merged**